tests/e2e/common/diagnostics/collector.go

```diff
@@ -0,0 +1,229 @@
+		return err
+	}
+	defer outputfile.Close()
+	if err = c.writeToFile(outputfile, objects); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createKubernetesClient(kubeconfig string) (*kubernetes.Clientset, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+type Collector struct {
+	*Config
+	collectors []DiagnosticCollector
+}
+type Option func(*Collector)
+
+func WithNamespace(namespace string) func(*Collector) {
+	return func(d *Collector) {
+		d.Config.namespace = namespace
+	}
+}
+
+func WithArtifactDir(artifactDir string) func(*Collector) {
+	return func(d *Collector) {
+		d.Config.artifactDir = artifactDir
+	}
+}
+
+func WithKubernetesClient(kubeconfig string) func(*Collector) {
+	return func(d *Collector) {
+		// Create the Kubernetes client
+		clientset, err := createKubernetesClient(kubeconfig)
+		if err != nil {
+			klog.ErrorS(err, "Error creating Kubernetes client")
+		}
+		d.clientset = clientset
+	}
+}
+
+func WichNFDClient(nfdClient *nfdclient.Clientset) func(*Collector) {
+	return func(d *Collector) {
+		d.NfdClient = nfdClient
+	}
+}
+
+func WithObjects(objects ...string) func(*Collector) {
+	// First let's remove duplicates
+	objectsMap := make(map[string]struct{})
+	for _, obj := range objects {
+		objectsMap[obj] = struct{}{}
+	}
+
+	return func(d *Collector) {
+		for obj := range objectsMap {
+			switch obj {
+			case Nodes:
+				d.collectors = append(d.collectors, nodes{Config: d.Config})
+			case Namespaces:
+				d.collectors = append(d.collectors, namespaces{Config: d.Config})
+			case Pods:
+				d.collectors = append(d.collectors, pods{Config: d.Config})
+			case Deployments:
+				d.collectors = append(d.collectors, deployments{Config: d.Config})
+			case DaemonSets:
+				d.collectors = append(d.collectors, daemonsets{Config: d.Config})
+			case Jobs:
+				d.collectors = append(d.collectors, jobs{Config: d.Config})
+			case NodeFeature:
+				d.collectors = append(d.collectors, nodeFeatures{Config: d.Config})
+			case NodeFeatureRule:
+				d.collectors = append(d.collectors, nodeFeatureRules{Config: d.Config})
+			}
+		}
+	}
+}
+
+func New(opts ...Option) (*Collector, error) {
+	c := &Config{}
+	dc := &Collector{
+		Config: c,
+	}
+
+	// use the variadic function to set the options
+	for _, opt := range opts {
+		opt(dc)
+	}
+
+	// Create the artifact directory
+	if err := os.MkdirAll(filepath.Join(c.artifactDir, c.namespace), os.ModePerm); err != nil {
+		return dc, err
+	}
+
+	return dc, nil
+}
+
+func (d *Collector) Collect(ctx context.Context) error {
+	var err error
+
+	// Redirect stdout and stderr to logs
+	logFile, err := d.createFile("diagnostic_collector.log")
+	if err != nil {
+		return err
+	}
+	defer logFile.Close()
+	d.log = logFile
+
+	// configure klog to write to the log file
+	klog.SetOutput(d.log)
```

Comment:
My question is whether we want to distinguish between output that is logged to the user and output that is logged for archiving. If we wanted different output for the user, the we would *not* override the global logger. If we wanted all output to end up in the file and not be visible to the tests, for example, then doing it as we do now is fine.

If we extend this to the `kubectl-nv` use case, there are probably some outputs that we want to output to `STDERR` instead of the configured file.

I'm happy to address this once that becomes a concern though.

Reply:
umm yup, klog takes loggr to enable custom configuration, we could create a MultiWritter to write to stdout and file

Reply:
I will do this one as a follow up, I want to do a PR to enhance the overall logging system of the E2E, I am not happy with how it its right now